lib/integrations/awsoidc/listec2.go

```diff
@@ -0,0 +1,175 @@
+
+	// Region is the AWS Region.
+	Region string
+
+	// NextToken is the token to be used to fetch the next page.
+	// If empty, the first page is fetched.
+	NextToken string
+}
+
+// CheckAndSetDefaults checks if the required fields are present.
+func (req *ListEC2Request) CheckAndSetDefaults() error {
+	if req.Integration == "" {
+		return trace.BadParameter("integration is required")
+	}
+	if req.Region == "" {
+		return trace.BadParameter("region is required")
+	}
+
+	return nil
+}
+
+// ListEC2Response contains a page of AWS EC2 Instances as Teleport Servers.
+type ListEC2Response struct {
+	// Servers contains the page of Servers.
+	Servers []types.Server
+
+	// NextToken is used for pagination.
+	// If non-empty, it can be used to request the next page.
+	NextToken string
+}
+
+// ListEC2Client describes the required methods to List EC2 Instances using a 3rd Party API.
+type ListEC2Client interface {
+	// Describes the specified instances or all instances.
+	DescribeInstances(ctx context.Context, params *ec2.DescribeInstancesInput, optFns ...func(*ec2.Options)) (*ec2.DescribeInstancesOutput, error)
+
+	// GetCallerIdentity returns details about the IAM user or role whose credentials are used to call the operation.
+	GetCallerIdentity(ctx context.Context, params *sts.GetCallerIdentityInput, optFns ...func(*sts.Options)) (*sts.GetCallerIdentityOutput, error)
+}
+
+type defaultListEC2Client struct {
+	*ec2.Client
+	stsClient *sts.Client
+}
+
+// GetCallerIdentity returns details about the IAM user or role whose credentials are used to call the operation.
+func (d defaultListEC2Client) GetCallerIdentity(ctx context.Context, params *sts.GetCallerIdentityInput, optFns ...func(*sts.Options)) (*sts.GetCallerIdentityOutput, error) {
+	return d.stsClient.GetCallerIdentity(ctx, params, optFns...)
+}
+
+// NewListEC2Client creates a new ListEC2Client using a AWSClientRequest.
+func NewListEC2Client(ctx context.Context, req *AWSClientRequest) (ListEC2Client, error) {
+	ec2Client, err := newEC2Client(ctx, req)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	stsClient, err := newSTSClient(ctx, req)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &defaultListEC2Client{
+		Client:    ec2Client,
+		stsClient: stsClient,
+	}, nil
+}
+
+// ListEC2 calls the following AWS API:
+// https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeInstances.html
+// It returns a list of EC2 Instances and an optional NextToken that can be used to fetch the next page
+func ListEC2(ctx context.Context, clt ListEC2Client, req ListEC2Request) (*ListEC2Response, error) {
+	if err := req.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	callerIdentity, err := clt.GetCallerIdentity(ctx, nil)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	accountID := aws.ToString(callerIdentity.Account)
+
+	describeEC2Instances := &ec2.DescribeInstancesInput{
+		Filters: []ec2Types.Filter{
+			filterRunningEC2Instance,
```

Comment:
Can we filter out instances to include just the distros we support?

Reply:
We support anything that supports the AWS API Calls: Send SSH Public Key and OpenTunnel.
We might even support more "exotic" distros or even other operating systems.

OTOH, I don't think we can filter out distros using the DescribeInstances EC2 API Call